File: src/math_parser/parser.cpp
#include <math_parser/parser.h>
#include <parser_exception/unexpected_char.h>
#include <iostream>
#include <vector>

Node* Parser::parse_expression() {
    Node* expression = this->parse_add_subtract();

    if(this->m_Tokenizer.get_token() != EOL) {
        throw UnexpectedCharException("Unexpected characters at end of expression.");
    }

    return expression;
}

Node* Parser::parse_add_subtract() {
    // parse left hand side
    Node* lhs = this->parse_multiply_divide();

    while(true) {
        const BinaryExecutor* operation;
        bool opFound = false;
        
        switch(this->m_Tokenizer.get_token()) {
            case Add:
                operation = &add_operation();
                opFound = true;
                break;

            case Subtract:
                operation = &subtract_operation();
                opFound = true;
                break;

            default:
                break;
        }

        // Binary operator not found
        if(!opFound) {
            return lhs;
        }

        // Skip the operator
        this->m_Tokenizer.next_token();

        // Parse the right hand side of the expression
        Node* rhs = this->parse_multiply_divide();

        // Create a binary node and use it as the left-hand side now on
        lhs = new BinaryNode(lhs, rhs, *operation);
    }
}

Node* Parser::parse_multiply_divide() {
    // parse left hand side
    Node* lhs = this->parse_unary();

    while(true) {
        const BinaryExecutor* operation;
        bool opFound = false;
        
        switch(this->m_Tokenizer.get_token()) {
            case Multiply:
                operation = &multiply_operation();
                opFound = true;
                break;

            case Divide:
                operation = &divide_operation();
                opFound = true;
                break;

            default:
                break;
        }

        // Binary operator not found
        if(!opFound) {
            return lhs;
        }

        // Skip the operator
        this->m_Tokenizer.next_token();

        // Parse the right hand side of the expression
        Node* rhs = this->parse_unary();

        // Create a binary node and use it as the left-hand side now on
        lhs = new BinaryNode(lhs, rhs, *operation);
    }
}

Node* Parser::parse_unary() {

    // Skip + because its no-op
    if(this->m_Tokenizer.get_token() == Add) {
        this->m_Tokenizer.next_token();
        return this->parse_unary();
    }

    // Negative operator
    if(this->m_Tokenizer.get_token() == Subtract){
        // Skip the token
        this->m_Tokenizer.next_token();

        // parse right side
        // this must recurse itself to support negative of negative
        Node* rhs = this->parse_unary();

        // Create the unary node
        return new UnaryNode(rhs, unary_operation());
    }

    // No positive/negative operator so parse a leaf node
    return this->parse_leaf();
}

Node* Parser::parse_leaf() {
    // Is it a number?
    if(this->m_Tokenizer.get_token() == Number) {
        Node* node = new NumberNode(this->m_Tokenizer.get_number());
        this->m_Tokenizer.next_token();
        return node;
    }

    // Is it a parenthesis?
    if(this->m_Tokenizer.get_token() == OpenParens) {
        // Skip the '('
        this->m_Tokenizer.next_token();

        // Parse the top-level expression
        Node* node = this->parse_add_subtract();

        // Check and skip ')'
        if(this->m_Tokenizer.get_token() != CloseParens) {
            throw UnexpectedCharException("Expected close parenthesis.");
        }

        this->m_Tokenizer.next_token();
        return node;
    }

    // Is it a variable or function?
    if(this->m_Tokenizer.get_token() == Identifier) {
        // capture the name and skip it
        std::string_view identifier = this->m_Tokenizer.get_current_identifier();
        this->m_Tokenizer.next_token();

        // Parenthesis indicate a function call, otherwise just a constant
        if(this->m_Tokenizer.get_token() != OpenParens) {
            return new ConstantNode(this->m_Tokenizer.get_current_identifier());
        } else {
            // Skip close parenthesis
            this->m_Tokenizer.next_token();

            std::vector<Node*> arguments;
            while(true) {
                // Parse argument and add to list
                arguments.emplace_back(this->parse_add_subtract());

                // Is there another argument?
                if(this->m_Tokenizer.get_token() == Comma) {
                    this->m_Tokenizer.next_token();
                    continue;
                }

                break;
            }

            // Check and skip ')'
            if(this->m_Tokenizer.get_token() != CloseParens) {
                throw UnexpectedCharException("Expected close parenthesis.");
            }

            this->m_Tokenizer.next_token();

            // Create the function node
            return new FunctionNode(identifier, arguments);
        }
    }
    
    throw UnexpectedCharException("Unexpected token: " + this->m_Tokenizer.get_token());
}

File: math_parser/parser.go
package math_parser

type Parser struct {
	tokenizer *Tokenizer
}

func NewParser(t *Tokenizer) *Parser {
	return &Parser{
		tokenizer: t,
	}
}

func (p *Parser) ParseExpression() Node {
	expression := p.parseAddSubtract()
	if p.tokenizer.currentToken != Token_EOL {
		panic("unexpected characters at end of expression")
	}

	return expression
}

func (p *Parser) parseAddSubtract() Node {
	lhs := p.parseMultiplyDivide()

	for {
		var operation *BinaryOperation = nil
		switch p.tokenizer.currentToken {
		case Token_Add:
			operation = &AddOperation

		case Token_Subtract:
			operation = &SubtractOperation
		}

		if operation == nil {
			return lhs
		}

		p.tokenizer.NextToken()

		rhs := p.parseMultiplyDivide()
		lhs = NewBinaryNode(lhs, rhs, operation)
	}
}

func (p *Parser) parseMultiplyDivide() Node {
	lhs := p.parseUnary()

	for {
		var operation *BinaryOperation = nil
		switch p.tokenizer.currentToken {
		case Token_Add:
			operation = &MultiplyOperation

		case Token_Subtract:
			operation = &DivideOperation
		}

		if operation == nil {
			return lhs
		}

		p.tokenizer.NextToken()

		rhs := p.parseUnary()
		lhs = NewBinaryNode(lhs, rhs, operation)
	}
}

func (p *Parser) parseUnary() Node {
	if p.tokenizer.currentToken == Token_Add {
		p.tokenizer.NextToken()
		return p.parseUnary()
	}

	if p.tokenizer.currentToken == Token_Subtract {
		p.tokenizer.NextToken()

		rhs := p.parseUnary()
		return NewUnaryNode(rhs, &NegativeOperation)
	}

	return p.parseLeaf()
}

func (p *Parser) parseLeaf() Node {
	if p.tokenizer.currentToken == Token_Number {
		node := NewNumberNode(p.tokenizer.currentNumber)
		p.tokenizer.NextToken()
		return node
	}

	if p.tokenizer.currentToken == Token_OpenParens {
		p.tokenizer.NextToken()

		node := p.parseAddSubtract()

		if p.tokenizer.currentToken != Token_CloseParens {
			panic("expected close parenthesis")
		}

		p.tokenizer.NextToken()
		return node
	}

	if p.tokenizer.currentToken == Token_Identifier {
		identifier := p.tokenizer.currentIdentifier

		p.tokenizer.NextToken()
		if p.tokenizer.currentToken == Token_OpenParens {
			return NewConstantNode(identifier)
		} else {
			p.tokenizer.NextToken()

			arguments := make([]Node, 0)
			for {
				arguments = append(arguments, p.parseAddSubtract())

				if p.tokenizer.currentToken == Token_Comma {
					p.tokenizer.NextToken()
					continue
				}

				break
			}

			if p.tokenizer.currentToken != Token_CloseParens {
				panic("expected close parenthesis")
			}

			p.tokenizer.NextToken()
			return NewFunctionNode(identifier, &arguments)
		}
	}

	panic("unexpected token " + string(p.tokenizer.currentToken))
}
